File: backend/app/schemas/analysis.py
"""
Analysis schemas for skill gap analysis requests and responses.

The AnalysisResponse is the main payload returned to the frontend
after an analysis completes. It contains scores, skills, and suggestions.
"""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.skill import ExtractedSkill, MissingSkill


class AnalysisRequest(BaseModel):
    """Input for creating a new analysis. Resume file is sent as multipart."""

    job_description: str = Field(min_length=50, max_length=10000)
    job_title: str | None = Field(None, max_length=500)
    job_company: str | None = Field(None, max_length=255)


class AnalysisSubmitResponse(BaseModel):
    """Returned immediately when an analysis job is submitted (async)."""

    job_id: UUID
    status: str = "queued"
    estimated_seconds: int = 15
    status_url: str
    ws_url: str


class AnalysisStatusResponse(BaseModel):
    """Real-time progress polling response."""

    job_id: UUID
    status: str  # queued | processing | completed | failed
    progress: int = Field(ge=0, le=100)
    current_step: str | None = None
    steps: list[dict] | None = None
    error_message: str | None = None


class ResumeSuggestion(BaseModel):
    """A single resume improvement suggestion."""

    section: str
    current: str
    suggested: str
    reason: str
    priority: str | None = None
    source: str | None = None


class CategoryBreakdownResponse(BaseModel):
    """Skill gap breakdown for a single category."""

    category: str
    display_name: str
    total_job_skills: int
    matched_count: int
    missing_count: int
    match_percentage: float
    matched_skills: list[str] = []
    missing_skills: list[str] = []
    priority: str


class ScoreExplanationResponse(BaseModel):
    """Human-readable score explanation."""

    match_score: float
    ats_score: float
    match_summary: str
    ats_summary: str
    strengths: list[str] = []
    weaknesses: list[str] = []
    overall_verdict: str


class ATSIssueResponse(BaseModel):
    """A single ATS compatibility issue."""

    severity: str
    category: str
    title: str
    description: str
    fix: str


class ATSCheckResponse(BaseModel):
    """ATS formatting check results."""

    issues: list[ATSIssueResponse] = []
    format_score: float
    passed_checks: int
    total_checks: int


class AnalysisResponse(BaseModel):
    """Full analysis result returned after completion."""

    id: UUID
    status: str
    match_score: float | None = Field(None, ge=0, le=100)
    ats_score: float | None = Field(None, ge=0, le=100)
    matched_skills: list[ExtractedSkill] = []
    missing_skills: list[MissingSkill] = []
    resume_skills: list[ExtractedSkill] = []
    job_skills: list[ExtractedSkill] = []
    suggestions: list[ResumeSuggestion] = []
    category_breakdowns: list[CategoryBreakdownResponse] = []
    score_explanation: ScoreExplanationResponse | None = None
    ats_check: ATSCheckResponse | None = None
    processing_time_ms: int | None = None
    ai_provider: str | None = None
    ai_model: str | None = None
    ai_tokens_used: int | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class AnalysisHistoryItem(BaseModel):
    """Summary item for the analysis history list."""

    id: UUID
    job_title: str | None
    job_company: str | None
    match_score: float | None
    ats_score: float | None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class AnalysisHistoryResponse(BaseModel):
    """Paginated list of past analyses."""

    analyses: list[AnalysisHistoryItem]
    total: int
    page: int
    per_page: int

File: frontend/src/types/analysis.ts
/**
 * TypeScript types for the analysis domain.
 *
 * These mirror the backend Pydantic schemas exactly.
 * Updated in Phase 8 to include Phase 6 gap analysis types.
 */

// ── Skill types ─────────────────────────────────────────────

export interface Skill {
  name: string;
  confidence: number; // 0.0 - 1.0
  category: string;
  source?: "resume" | "job_description";
}

export interface MissingSkill {
  name: string;
  priority: "high" | "medium" | "low";
  category: string;
  weight?: number;
}

// ── Resume suggestion ───────────────────────────────────────

export interface ResumeSuggestion {
  section: string;
  current: string;
  suggested: string;
  reason: string;
  priority?: "high" | "medium" | "low";
  source?: "rule" | "llm";
}

// ── Category breakdown (Phase 6) ────────────────────────────

export interface CategoryBreakdown {
  category: string;
  display_name: string;
  total_job_skills: number;
  matched_count: number;
  missing_count: number;
  match_percentage: number;
  matched_skills: string[];
  missing_skills: string[];
  priority: "critical" | "important" | "nice_to_have";
}

// ── Score explanation (Phase 6) ─────────────────────────────

export interface ScoreExplanation {
  match_score: number;
  ats_score: number;
  match_summary: string;
  ats_summary: string;
  strengths: string[];
  weaknesses: string[];
  overall_verdict: string;
}

// ── ATS check (Phase 6) ────────────────────────────────────

export interface ATSIssue {
  severity: "error" | "warning" | "info";
  category: string;
  title: string;
  description: string;
  fix: string;
}

export interface ATSCheck {
  issues: ATSIssue[];
  format_score: number;
  passed_checks: number;
  total_checks: number;
}

// ── Analysis result ─────────────────────────────────────────

export interface AnalysisResult {
  id: string;
  status: "queued" | "processing" | "completed" | "failed";
  match_score: number | null;
  ats_score: number | null;
  matched_skills: Skill[];
  missing_skills: MissingSkill[];
  resume_skills: Skill[];
  job_skills: Skill[];
  suggestions: ResumeSuggestion[];
  category_breakdowns: CategoryBreakdown[];
  score_explanation: ScoreExplanation | null;
  ats_check: ATSCheck | null;
  processing_time_ms: number | null;
  ai_provider: string | null;
  ai_model: string | null;
  ai_tokens_used: number | null;
  suggestions_limited?: boolean;
  created_at: string;
}

// ── Async job types ─────────────────────────────────────────

export interface AnalysisSubmitResponse {
  job_id: string;
  status: string;
  estimated_seconds: number;
  status_url: string;
  ws_url: string;
}

export interface AnalysisStatusResponse {
  job_id: string;
  status: "queued" | "processing" | "completed" | "failed";
  progress: number; // 0-100
  current_step: string | null;
  error_message: string | null;
}

// ── History ─────────────────────────────────────────────────

export interface AnalysisHistoryItem {
  id: string;
  job_title: string | null;
  job_company: string | null;
  match_score: number | null;
  ats_score: number | null;
  status: string;
  created_at: string;
}

// ── Roadmap (Phase 9) ──────────────────────────────────────

export interface RoadmapPhase {
  week_range: string;
  focus: string;
  objectives: string[];
  resources: string[];
}

export interface RoadmapResponse {
  id: string;
  analysis_id: string;
  total_weeks: number;
  phases: RoadmapPhase[];
}

// ── Resume Advisor (Phase 9) ───────────────────────────────

export interface SectionRewrite {
  section: string;
  original: string;
  rewritten: string;
  changes_made: string[];
  confidence: number;
}

export interface AdvisorResponse {
  rewrites: SectionRewrite[];
  overall_summary: string;
}

// ── Resume types ────────────────────────────────────────────

export interface ResumeUploadResponse {
  id: string;
  original_filename: string;
  file_type: string;
  file_size_bytes: number;
  created_at: string;
  last_used_at: string | null;
}

export interface PaginatedResumeResponse {
  resumes: ResumeUploadResponse[];
  total: number;
  skip: number;
  limit: number;
}
